src/main/java/com/zufar/icedlatte/auth/google/endpoint/AuthEndpoint.java

```diff
@@ -0,0 +1,189 @@
+            + "/callback";
+    String url =
+        authorizationServerUrl
+            + "?"
+            + "scope="
+            + scope
+            + "&"
+            + "access_type=offline&"
+            + "include_granted_scopes=true&"
+            + "response_type=code&"
+            + "state=state_parameter_passthrough_value&"
+            + "redirect_uri="
+            + redirectAuthUri
+            + "&"
+            + "client_id="
+            + clientId;
+    return ResponseEntity.status(HttpStatus.TEMPORARY_REDIRECT).header("Location", url).build();
+  }
+
+  @GetMapping("/callback")
+  public ResponseEntity<?> googleAuthCallback(@RequestParam("code") String code)
+      throws IOException, GeneralSecurityException, InterruptedException {
+    log.warn("Received callback the request a google auth");
+
+    // get token
+    AuthorizationCodeFlow flow = initializeFlow();
+    String redirectAuthUri =
+        "http://localhost"
+            + ":"
+            + environment.getProperty("server.port")
+            + GOOGLE_AUTH
+            + "/callback";
+    TokenResponse token = flow.newTokenRequest(code).setRedirectUri(redirectAuthUri).execute();
+
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = GsonFactory.getDefaultInstance();
+    GoogleIdTokenVerifier verifier =
+        new GoogleIdTokenVerifier.Builder(httpTransport, jsonFactory)
+            .setAudience(Collections.singletonList(clientId))
+            .build();
+
+    // verify idToken
+    GoogleIdToken idToken = verifier.verify((String) token.get("id_token"));
+    if (idToken != null) {
+      GoogleIdToken.Payload payload = idToken.getPayload();
+      String email = payload.getEmail();
+      log.info("User email: {}", email);
+
+      // find or create user
+      String password = findOrCreateUser(email);
+      log.info("Success logged with email={}", email);
+
+      // login in main form
+      sendPostRequest(authenticateUri, new UserAuthenticationRequest(email, password));
+
+      // redirect on main page
+      return ResponseEntity.status(HttpStatus.TEMPORARY_REDIRECT)
+          .header("Location", redirectUri)
+          .build();
+    } else {
+      log.error("Invalid ID token.");
+      return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+    }
+  }
+
+  private GoogleAuthorizationCodeFlow initializeFlow()
+      throws IOException, GeneralSecurityException {
+    HttpTransport httpTransport = GoogleNetHttpTransport.newTrustedTransport();
+    JsonFactory jsonFactory = GsonFactory.getDefaultInstance();
+    return new GoogleAuthorizationCodeFlow.Builder(
+            httpTransport, jsonFactory, clientId, clientSecret, List.of(scope))
+        .setAccessType("offline")
+        .setApprovalPrompt("force")
+        .build();
+  }
+
+  private String findOrCreateUser(String email) {
+    Optional<UserEntity> userEntity = userCrudRepository.findByEmail(email);
```

Comment:
I'm not sure, but apparently this should be annotated with `@Transactional(readOnly = true)`. Also, I would extract this line to a separate method in `SingleUserProvider`.

Reply:
dont want try-catch

Reply:
extract in SingleUserProvider